File: src/brs/db/sql/SqlAccountStore.java
package brs.db.sql;

import brs.Account;
import brs.Account.AccountAsset;
import brs.Asset;
import brs.Signum;
import brs.Transaction;
import brs.TransactionType;
import brs.db.VersionedBatchEntityTable;
import brs.db.VersionedEntityTable;
import brs.db.cache.DBCacheManagerImpl;
import brs.db.store.AccountStore;
import brs.db.store.DerivedTableManager;
import brs.fluxcapacitor.FluxValues;
import brs.props.Props;
import brs.schema.tables.records.AccountBalanceRecord;
import brs.util.Convert;
import signumj.crypto.SignumCrypto;

import org.jooq.*;
import org.jooq.Record;
import org.jooq.impl.DSL;
import org.slf4j.LoggerFactory;

import java.math.BigDecimal;
import java.util.*;

import static brs.schema.Tables.*;

public class SqlAccountStore implements AccountStore {

  private static final DbKey.LongKeyFactory<Account> accountDbKeyFactory = new DbKey.LongKeyFactory<Account>(ACCOUNT.ID) {
    @Override
    public DbKey newKey(Account account) {
      return (DbKey) account.nxtKey;
    }
  };
  private static final DbKey.LongKeyFactory<Account.Balance> accountBalanceDbKeyFactory = new DbKey.LongKeyFactory<Account.Balance>(ACCOUNT_BALANCE.ID) {
    @Override
    public DbKey newKey(Account.Balance account) {
      return (DbKey) account.nxtKey;
    }
  };
  private static final DbKey.LongKeyFactory<Account.RewardRecipientAssignment> rewardRecipientAssignmentDbKeyFactory
    = new DbKey.LongKeyFactory<Account.RewardRecipientAssignment>(REWARD_RECIP_ASSIGN.ACCOUNT_ID) {
    @Override
    public DbKey newKey(Account.RewardRecipientAssignment assignment) {
      return (DbKey) assignment.signumKey;
    }
  };
  private static final org.slf4j.Logger logger = LoggerFactory.getLogger(SqlAccountStore.class);
  private static final DbKey.LinkKeyFactory<Account.AccountAsset> accountAssetDbKeyFactory
    = new DbKey.LinkKeyFactory<Account.AccountAsset>("account_id", "asset_id") {
    @Override
    public DbKey newKey(Account.AccountAsset accountAsset) {
      return (DbKey) accountAsset.signumKey;
    }
  };

  private static final Integer InsertMaxBatchSize = Signum.getPropertyService().getInt(Props.DB_INSERT_BATCH_MAX_SIZE);

  private static final Set<String> PK_CHECKS = Collections
    .unmodifiableSet(new HashSet<>(Signum.getPropertyService().getStringList(Props.BRS_PK_CHECKS)));

  public SqlAccountStore(DerivedTableManager derivedTableManager, DBCacheManagerImpl dbCacheManager) {
    rewardRecipientAssignmentTable = new VersionedEntitySqlTable<Account.RewardRecipientAssignment>("reward_recip_assign", brs.schema.Tables.REWARD_RECIP_ASSIGN, rewardRecipientAssignmentDbKeyFactory, derivedTableManager) {

      @Override
      protected Account.RewardRecipientAssignment load(DSLContext ctx, Record rs) {
        return new SqlRewardRecipientAssignment(rs);
      }

      @Override
      protected void save(DSLContext ctx, Account.RewardRecipientAssignment assignment) {

        ctx.insertInto(REWARD_RECIP_ASSIGN,
            REWARD_RECIP_ASSIGN.ACCOUNT_ID, REWARD_RECIP_ASSIGN.PREV_RECIP_ID,
            REWARD_RECIP_ASSIGN.RECIP_ID, REWARD_RECIP_ASSIGN.FROM_HEIGHT,
            REWARD_RECIP_ASSIGN.HEIGHT, REWARD_RECIP_ASSIGN.LATEST)
          .values(assignment.getAccountId(), assignment.getPrevRecipientId(),
            assignment.getRecipientId(), assignment.getFromHeight(),
            Signum.getBlockchain().getHeight(), true)
          .onConflict(REWARD_RECIP_ASSIGN.ACCOUNT_ID, REWARD_RECIP_ASSIGN.HEIGHT)
          .doUpdate()
          .set(REWARD_RECIP_ASSIGN.PREV_RECIP_ID, assignment.getPrevRecipientId())
          .set(REWARD_RECIP_ASSIGN.RECIP_ID, assignment.getRecipientId())
          .set(REWARD_RECIP_ASSIGN.FROM_HEIGHT, assignment.getFromHeight())
          .set(REWARD_RECIP_ASSIGN.LATEST, true)
          .execute();

      }
    };

    accountAssetTable = new VersionedEntitySqlTable<Account.AccountAsset>("account_asset", brs.schema.Tables.ACCOUNT_ASSET, accountAssetDbKeyFactory, derivedTableManager) {
      private final List<SortField<?>> sort = initializeSort();

      private List<SortField<?>> initializeSort() {
        List<SortField<?>> sort = new ArrayList<>();
        sort.add(tableClass.field("quantity", Long.class).desc());
        sort.add(tableClass.field("account_id", Long.class).asc());
        sort.add(tableClass.field("asset_id", Long.class).asc());
        return Collections.unmodifiableList(sort);
      }

      @Override
      protected Account.AccountAsset load(DSLContext ctx, Record rs) {
        return new SQLAccountAsset(rs);
      }

      @Override
      protected void save(DSLContext ctx, Account.AccountAsset accountAsset) {
        ctx.insertInto(ACCOUNT_ASSET,
            ACCOUNT_ASSET.ACCOUNT_ID, ACCOUNT_ASSET.ASSET_ID,
            ACCOUNT_ASSET.QUANTITY, ACCOUNT_ASSET.UNCONFIRMED_QUANTITY,
            ACCOUNT_ASSET.HEIGHT, ACCOUNT_ASSET.LATEST)
          .values(accountAsset.getAccountId(), accountAsset.getAssetId(),
            accountAsset.getQuantityQnt(), accountAsset.getUnconfirmedQuantityQnt(),
            Signum.getBlockchain().getHeight(), true)
          .onConflict(ACCOUNT_ASSET.ACCOUNT_ID, ACCOUNT_ASSET.ASSET_ID, ACCOUNT_ASSET.HEIGHT)
          .doUpdate()
          .set(ACCOUNT_ASSET.QUANTITY, accountAsset.getQuantityQnt())
          .set(ACCOUNT_ASSET.UNCONFIRMED_QUANTITY, accountAsset.getUnconfirmedQuantityQnt())
          .set(ACCOUNT_ASSET.LATEST, true)
          .execute();

      }

      @Override
      protected List<SortField<?>> defaultSort() {
        return sort;
      }
    };

    accountTable = new VersionedBatchEntitySqlTable<Account>("account", brs.schema.Tables.ACCOUNT, accountDbKeyFactory, derivedTableManager, dbCacheManager, Account.class) {
      @Override
      protected Account load(DSLContext ctx, Record rs) {
        return new SqlAccount(rs);
      }

      @Override
      protected void bulkInsert(DSLContext ctx, Collection<Account> accounts) {
        List<Query> accountQueries = new ArrayList<>();
        int height = Signum.getBlockchain().getHeight();

        for (Account account : accounts) {
          if (account == null) continue;
          accountQueries.add(
            ctx.insertInto(ACCOUNT)
              .columns(ACCOUNT.ID, ACCOUNT.HEIGHT, ACCOUNT.CREATION_HEIGHT, ACCOUNT.PUBLIC_KEY, ACCOUNT.KEY_HEIGHT,
                ACCOUNT.NAME, ACCOUNT.DESCRIPTION, ACCOUNT.LATEST)
              .values(account.getId(), height, account.getCreationHeight(), account.getPublicKey(), account.getKeyHeight(),
                account.getName(), account.getDescription(), true)
              .onConflict(ACCOUNT.ID, ACCOUNT.HEIGHT)
              .doUpdate()
              .set(ACCOUNT.CREATION_HEIGHT, account.getCreationHeight())
              .set(ACCOUNT.PUBLIC_KEY, account.getPublicKey())
              .set(ACCOUNT.KEY_HEIGHT, account.getKeyHeight())
              .set(ACCOUNT.NAME, account.getName())
              .set(ACCOUNT.DESCRIPTION, account.getDescription())
              .set(ACCOUNT.LATEST, true)
          );
        }
        ctx.batch(accountQueries).execute();
      }
    };

    accountBalanceTable = new VersionedBatchEntitySqlTable<Account.Balance>("account_balance", brs.schema.Tables.ACCOUNT_BALANCE, accountBalanceDbKeyFactory, derivedTableManager, dbCacheManager, Account.Balance.class) {
      @Override
      protected Account.Balance load(DSLContext ctx, Record rs) {
        return new SqlAccountBalance(rs);
      }

      @Override
      protected void bulkInsert(DSLContext ctx, Collection<Account.Balance> accounts) {
        int height = Signum.getBlockchain().getHeight();
        Iterator<Account.Balance> iterator = accounts.iterator();
        List<Record6<Long, Integer, Long, Long, Long, Boolean>> rows = new ArrayList<>();
        while (iterator.hasNext()) {
          Account.Balance balance = iterator.next();
          if(balance == null) {
            continue;
          }

          rows.add(ctx.newRecord(ACCOUNT_BALANCE.ID, ACCOUNT_BALANCE.HEIGHT,
              ACCOUNT_BALANCE.BALANCE, ACCOUNT_BALANCE.UNCONFIRMED_BALANCE,
              ACCOUNT_BALANCE.FORGED_BALANCE, ACCOUNT.LATEST)
            .values(balance.getId(), height,
              balance.getBalanceNqt(), balance.getUnconfirmedBalanceNqt(),
              balance.getForgedBalanceNqt(), true));

          if(rows.size() >= InsertMaxBatchSize){
            ctx.insertInto(ACCOUNT_BALANCE, ACCOUNT_BALANCE.ID, ACCOUNT_BALANCE.HEIGHT,
                ACCOUNT_BALANCE.BALANCE, ACCOUNT_BALANCE.UNCONFIRMED_BALANCE,
                ACCOUNT_BALANCE.FORGED_BALANCE, ACCOUNT.LATEST)
              .valuesOfRecords(rows)
              .execute();
            rows.clear();
          }
        }

        if(!rows.isEmpty()){
          ctx.insertInto(ACCOUNT_BALANCE, ACCOUNT_BALANCE.ID, ACCOUNT_BALANCE.HEIGHT,
              ACCOUNT_BALANCE.BALANCE, ACCOUNT_BALANCE.UNCONFIRMED_BALANCE,
              ACCOUNT_BALANCE.FORGED_BALANCE, ACCOUNT.LATEST)
            .valuesOfRecords(rows)
            .execute();
        }
      }
    };
  }

  private static Condition getAccountsWithRewardRecipientClause(final long id, final int height) {
    return REWARD_RECIP_ASSIGN.RECIP_ID.eq(id).and(REWARD_RECIP_ASSIGN.FROM_HEIGHT.le(height));
  }

  private final VersionedEntityTable<Account.AccountAsset> accountAssetTable;

  private final VersionedEntityTable<Account.RewardRecipientAssignment> rewardRecipientAssignmentTable;

  private final VersionedBatchEntityTable<Account> accountTable;

  private final VersionedBatchEntityTable<Account.Balance> accountBalanceTable;

  @Override
  public VersionedBatchEntityTable<Account> getAccountTable() {
    return accountTable;
  }

  @Override
  public VersionedBatchEntityTable<Account.Balance> getAccountBalanceTable() {
    return accountBalanceTable;
  }


  @Override
  public VersionedEntityTable<Account.RewardRecipientAssignment> getRewardRecipientAssignmentTable() {
    return rewardRecipientAssignmentTable;
  }

  @Override
  public DbKey.LongKeyFactory<Account.RewardRecipientAssignment> getRewardRecipientAssignmentKeyFactory() {
    return rewardRecipientAssignmentDbKeyFactory;
  }

  @Override
  public DbKey.LinkKeyFactory<Account.AccountAsset> getAccountAssetKeyFactory() {
    return accountAssetDbKeyFactory;
  }

  @Override
  public VersionedEntityTable<Account.AccountAsset> getAccountAssetTable() {
    return accountAssetTable;
  }

  @Override
  public long getAllAccountsBalance() {
    return Db.useDSLContext(ctx -> {
      return ctx.select(DSL.sum(ACCOUNT_BALANCE.BALANCE)).from(ACCOUNT_BALANCE).where(ACCOUNT_BALANCE.LATEST.isTrue())
        .fetchOneInto(long.class);
    });
  }

  @Override
  public int getAssetAccountsCount(Asset asset, long minimumQuantity, boolean ignoreTreasury, boolean unconfirmed) {
    return Db.useDSLContext(ctx -> {

      SelectConditionStep<Record1<Integer>> select = ctx.selectCount().from(ACCOUNT_ASSET)
        .where(ACCOUNT_ASSET.ASSET_ID.eq(asset.getId())).and(ACCOUNT_ASSET.LATEST.isTrue())
        .and(ACCOUNT_ASSET.ACCOUNT_ID.ne(0L));
      if (minimumQuantity > 0L) {
        select = select.and((unconfirmed ? ACCOUNT_ASSET.UNCONFIRMED_QUANTITY : ACCOUNT_ASSET.QUANTITY).ge(minimumQuantity));
      }
      if (ignoreTreasury) {
        Transaction transaction = Signum.getBlockchain().getTransaction(asset.getId());
        if (transaction != null) {
          List<Long> ignoredAccounts = ctx.select(TRANSACTION.RECIPIENT_ID).from(TRANSACTION)
            .where(TRANSACTION.TYPE.eq(TransactionType.TYPE_COLORED_COINS.getType()))
            .and(TRANSACTION.SUBTYPE.eq(TransactionType.SUBTYPE_COLORED_COINS_ADD_TREASURY_ACCOUNT))
            .and(TRANSACTION.REFERENCED_TRANSACTION_FULLHASH.eq(Convert.parseHexString(transaction.getFullHash())))
            .fetch().getValues(TRANSACTION.RECIPIENT_ID);
          select = select.and(ACCOUNT_ASSET.ACCOUNT_ID.notIn(ignoredAccounts));
        }
      }
      return select.fetchOne(0, int.class);
    });
  }

  @Override
  public long getAssetCirculatingSupply(Asset asset, boolean ignoreTreasury, boolean unconfirmed) {
    return Db.useDSLContext(ctx -> {

      SelectConditionStep<Record1<BigDecimal>> select = ctx.select(DSL.sum(
          unconfirmed ? ACCOUNT_ASSET.UNCONFIRMED_QUANTITY : ACCOUNT_ASSET.QUANTITY))
        .from(ACCOUNT_ASSET).where(ACCOUNT_ASSET.ASSET_ID.eq(asset.getId()))
        .and(ACCOUNT_ASSET.LATEST.isTrue())
        .and(ACCOUNT_ASSET.ACCOUNT_ID.ne(0L));

      if (ignoreTreasury) {
        Transaction transaction = Signum.getBlockchain().getTransaction(asset.getId());
        if (transaction != null) {
          List<Long> ignoredAccounts = ctx.select(TRANSACTION.RECIPIENT_ID).from(TRANSACTION)
            .where(TRANSACTION.TYPE.eq(TransactionType.TYPE_COLORED_COINS.getType()))
            .and(TRANSACTION.SUBTYPE.eq(TransactionType.SUBTYPE_COLORED_COINS_ADD_TREASURY_ACCOUNT))
            .and(TRANSACTION.REFERENCED_TRANSACTION_FULLHASH.eq(Convert.parseHexString(transaction.getFullHash())))
            .fetch().getValues(TRANSACTION.RECIPIENT_ID);
          select = select.and(ACCOUNT_ASSET.ACCOUNT_ID.notIn(ignoredAccounts));
        }
      }

      return select.fetchOne(0, long.class);
    });
  }

  @Override
  public DbKey.LongKeyFactory<Account> getAccountKeyFactory() {
    return accountDbKeyFactory;
  }

  @Override
  public DbKey.LongKeyFactory<Account.Balance> getAccountBalanceKeyFactory() {
    return accountBalanceDbKeyFactory;
  }

  @Override
  public Collection<Account.RewardRecipientAssignment> getAccountsWithRewardRecipient(Long recipientId) {
    return getRewardRecipientAssignmentTable().getManyBy(getAccountsWithRewardRecipientClause(recipientId, Signum.getBlockchain().getHeight() + 1), 0, -1);
  }

  @Override
  public Collection<Account.AccountAsset> getAssets(int from, int to, Long id) {
    return getAccountAssetTable().getManyBy(ACCOUNT_ASSET.ACCOUNT_ID.eq(id), from, to);
  }

  @Override
  public Account.AccountAsset getAccountAsset(Long accountId, Long assetId) {
    return getAccountAssetTable().getBy(ACCOUNT_ASSET.ACCOUNT_ID.eq(accountId).and(ACCOUNT_ASSET.ASSET_ID.eq(assetId)));
  }

  @Override
  public Collection<Account.AccountAsset> getAssetAccounts(Asset asset, boolean ignoreTreasury, long minimumQuantity, boolean unconfirmed, int from, int to) {
    List<SortField<?>> sort = new ArrayList<>();
    sort.add(ACCOUNT_ASSET.field("quantity", Long.class).desc());
    sort.add(ACCOUNT_ASSET.field("account_id", Long.class).asc());

    Condition condition = ACCOUNT_ASSET.ASSET_ID.eq(asset.getId());
    if (minimumQuantity > 0L) {
      condition = condition.and((unconfirmed ? ACCOUNT_ASSET.UNCONFIRMED_QUANTITY : ACCOUNT_ASSET.QUANTITY).ge(minimumQuantity));
    }
    ArrayList<Long> treasuryAccounts = new ArrayList<>();
    // the 0 account should also be removed from the circulating
    treasuryAccounts.add(0L);
    Transaction transaction = Signum.getBlockchain().getTransaction(asset.getId());
    if (transaction != null) {
      treasuryAccounts.addAll(Db.useDSLContext(ctx -> {
        return ctx.select(TRANSACTION.RECIPIENT_ID).from(TRANSACTION).where(TRANSACTION.TYPE.eq(TransactionType.TYPE_COLORED_COINS.getType()))
          .and(TRANSACTION.SUBTYPE.eq(TransactionType.SUBTYPE_COLORED_COINS_ADD_TREASURY_ACCOUNT))
          .and(TRANSACTION.REFERENCED_TRANSACTION_FULLHASH.eq(Convert.parseHexString(transaction.getFullHash())))
          .fetch().getValues(TRANSACTION.RECIPIENT_ID);
      }));
    }
    if (ignoreTreasury) {
      condition = condition.and(ACCOUNT_ASSET.ACCOUNT_ID.notIn(treasuryAccounts));
    }
    Collection<AccountAsset> accounts = getAccountAssetTable().getManyBy(condition, from, to, sort);

    // flag treasury accounts
    for (AccountAsset account : accounts) {
      if (treasuryAccounts.contains(account.getAccountId())) {
        account.setTreasury(true);
      }
    }

    return accounts;
  }

  @Override
  public boolean setOrVerify(Account acc, byte[] key, int height) {
    if (acc.getPublicKey() == null) {
      if (Signum.getFluxCapacitor().getValue(FluxValues.PK_FREEZE)
        && Signum.getBlockchain().getHeight() - acc.getCreationHeight() > Signum.getPropertyService().getInt(Props.PK_BLOCKS_PAST)) {
        logger.info("Setting a new key for an old account {} is not allowed, created at height {}", Convert.toUnsignedLong(acc.id), acc.getCreationHeight());
        return false;
      }

      if (Db.isInTransaction()) {
        acc.setPublicKey(key);
        acc.setKeyHeight(-1);
        getAccountTable().insert(acc);
      }
      return true;
    } else if (Signum.getFluxCapacitor().getValue(FluxValues.PK_FREEZE)
      && PK_CHECKS.contains(Convert.toHexString(SignumCrypto.getInstance().longToBytesLE(acc.getId())))) {
      logger.info("Using the key for account {}", Convert.toUnsignedLong(acc.id));
      return false;
    } else if (Arrays.equals(acc.getPublicKey(), key)) {
      return true;
    } else if (acc.getKeyHeight() == -1) {
      if (logger.isInfoEnabled()) {
        logger.info("DUPLICATE KEY!!!");
        logger.info("Account key for {} was already set to a different one at the same height, current height is {}, rejecting new key", Convert.toUnsignedLong(acc.id), height);
      }
      return false;
    } else if (acc.getKeyHeight() >= height) {
      logger.info("DUPLICATE KEY!!!");
      if (Db.isInTransaction()) {
        if (logger.isInfoEnabled()) {
          logger.info("Changing key for account {} at height {}, was previously set to a different one at height {}", Convert.toUnsignedLong(acc.id), height, acc.getKeyHeight());
        }
        acc.setPublicKey(key);
        acc.setKeyHeight(height);
        getAccountTable().insert(acc);
      }
      return true;
    }
    if (logger.isInfoEnabled()) {
      logger.info("DUPLICATE KEY!!!");
      logger.info("Invalid key for account {} at height {}, was already set to a different one at height {}", Convert.toUnsignedLong(acc.id), height, acc.getKeyHeight());
    }
    return false;
  }

  static class SQLAccountAsset extends Account.AccountAsset {
    SQLAccountAsset(Record rs) {
      super(rs.get(ACCOUNT_ASSET.ACCOUNT_ID),
        rs.get(ACCOUNT_ASSET.ASSET_ID),
        rs.get(ACCOUNT_ASSET.QUANTITY),
        rs.get(ACCOUNT_ASSET.UNCONFIRMED_QUANTITY),
        accountAssetDbKeyFactory.newKey(rs.get(ACCOUNT_ASSET.ACCOUNT_ID), rs.get(ACCOUNT_ASSET.ASSET_ID))
      );
    }
  }

  class SqlAccount extends Account {
    SqlAccount(Record record) {
      super(record.get(ACCOUNT.ID), accountDbKeyFactory.newKey(record.get(ACCOUNT.ID)),
        record.get(ACCOUNT.CREATION_HEIGHT));
      this.setPublicKey(record.get(ACCOUNT.PUBLIC_KEY));
      this.setKeyHeight(record.get(ACCOUNT.KEY_HEIGHT));
      this.name = record.get(ACCOUNT.NAME);
      this.description = record.get(ACCOUNT.DESCRIPTION);
    }
  }

  class SqlAccountBalance extends Account.Balance {
    SqlAccountBalance(Record record) {
      super(record.get(ACCOUNT_BALANCE.ID));
      this.balanceNqt = record.get(ACCOUNT_BALANCE.BALANCE);
      this.unconfirmedBalanceNqt = record.get(ACCOUNT_BALANCE.UNCONFIRMED_BALANCE);
      this.forgedBalanceNqt = record.get(ACCOUNT_BALANCE.FORGED_BALANCE);
    }
  }

  class SqlRewardRecipientAssignment extends Account.RewardRecipientAssignment {
    SqlRewardRecipientAssignment(Record record) {
      super(
        record.get(REWARD_RECIP_ASSIGN.ACCOUNT_ID),
        record.get(REWARD_RECIP_ASSIGN.PREV_RECIP_ID),
        record.get(REWARD_RECIP_ASSIGN.RECIP_ID),
        record.get(REWARD_RECIP_ASSIGN.FROM_HEIGHT),
        rewardRecipientAssignmentDbKeyFactory.newKey(record.get(REWARD_RECIP_ASSIGN.ACCOUNT_ID))
      );
    }
  }


}

File: src/main/java/brs/db/sql/SqlAccountStore.kt
package brs.db.sql

import brs.Account
import brs.DependencyProvider
import brs.db.VersionedBatchEntityTable
import brs.db.VersionedEntityTable
import brs.db.store.AccountStore
import brs.schema.Tables.*
import brs.schema.tables.records.AccountAssetRecord
import brs.schema.tables.records.AccountRecord
import brs.schema.tables.records.RewardRecipAssignRecord
import brs.util.toUnsignedString
import org.jooq.*
import org.slf4j.LoggerFactory
import java.util.*

class SqlAccountStore(private val dp: DependencyProvider) : AccountStore {

    override val accountAssetTable: VersionedEntityTable<Account.AccountAsset>

    override val rewardRecipientAssignmentTable: VersionedEntityTable<Account.RewardRecipientAssignment>

    override val accountTable: VersionedBatchEntityTable<Account>

    override val rewardRecipientAssignmentKeyFactory: DbKey.LongKeyFactory<Account.RewardRecipientAssignment>
        get() = rewardRecipientAssignmentDbKeyFactory

    override val accountAssetKeyFactory: DbKey.LinkKeyFactory<Account.AccountAsset>
        get() = accountAssetDbKeyFactory

    override val accountKeyFactory: DbKey.LongKeyFactory<Account>
        get() = accountDbKeyFactory

    init {
        rewardRecipientAssignmentTable = object : VersionedEntitySqlTable<Account.RewardRecipientAssignment>("reward_recip_assign", brs.schema.Tables.REWARD_RECIP_ASSIGN, rewardRecipientAssignmentDbKeyFactory, dp) {

            override fun load(ctx: DSLContext, rs: Record): Account.RewardRecipientAssignment {
                return SqlRewardRecipientAssignment(rs)
            }

            override fun save(ctx: DSLContext, assignment: Account.RewardRecipientAssignment) {
                ctx.mergeInto<RewardRecipAssignRecord, Long, Long, Long, Int, Int, Boolean>(REWARD_RECIP_ASSIGN, REWARD_RECIP_ASSIGN.ACCOUNT_ID, REWARD_RECIP_ASSIGN.PREV_RECIP_ID, REWARD_RECIP_ASSIGN.RECIP_ID, REWARD_RECIP_ASSIGN.FROM_HEIGHT, REWARD_RECIP_ASSIGN.HEIGHT, REWARD_RECIP_ASSIGN.LATEST)
                        .key(REWARD_RECIP_ASSIGN.ACCOUNT_ID, REWARD_RECIP_ASSIGN.HEIGHT)
                        .values(assignment.accountId, assignment.prevRecipientId, assignment.recipientId, assignment.fromHeight, dp.blockchain.height, true)
                        .execute()
            }
        }

        accountAssetTable = object : VersionedEntitySqlTable<Account.AccountAsset>("account_asset", brs.schema.Tables.ACCOUNT_ASSET, accountAssetDbKeyFactory, dp) {
            private val sort = initializeSort()

            private fun initializeSort(): List<SortField<*>> {
                val sort = mutableListOf<SortField<*>>()
                sort.add(tableClass.field("quantity", Long::class.java).desc())
                sort.add(tableClass.field("account_id", Long::class.java).asc())
                sort.add(tableClass.field("asset_id", Long::class.java).asc())
                return sort
            }

            override fun load(ctx: DSLContext, rs: Record): Account.AccountAsset {
                return SQLAccountAsset(rs)
            }

            override fun save(ctx: DSLContext, accountAsset: Account.AccountAsset) {
                ctx.mergeInto<AccountAssetRecord, Long, Long, Long, Long, Int, Boolean>(ACCOUNT_ASSET, ACCOUNT_ASSET.ACCOUNT_ID, ACCOUNT_ASSET.ASSET_ID, ACCOUNT_ASSET.QUANTITY, ACCOUNT_ASSET.UNCONFIRMED_QUANTITY, ACCOUNT_ASSET.HEIGHT, ACCOUNT_ASSET.LATEST)
                        .key(ACCOUNT_ASSET.ACCOUNT_ID, ACCOUNT_ASSET.ASSET_ID, ACCOUNT_ASSET.HEIGHT)
                        .values(accountAsset.accountId, accountAsset.assetId, accountAsset.quantityQNT, accountAsset.unconfirmedQuantityQNT, dp.blockchain.height, true)
                        .execute()
            }

            override fun defaultSort(): List<SortField<*>> {
                return sort
            }
        }

        accountTable = object : VersionedBatchEntitySqlTable<Account>("account", brs.schema.Tables.ACCOUNT, accountDbKeyFactory, Account::class.java, dp) {
            override fun load(ctx: DSLContext, rs: Record): Account {
                return SqlAccount(rs)
            }

            override fun bulkInsert(ctx: DSLContext, accounts: Collection<Account>) {
                val accountQueries = mutableListOf<Query>()
                val height = dp.blockchain.height
                for (account in accounts) {
                    accountQueries.add(
                            ctx.mergeInto<AccountRecord, Long, Int, Int, ByteArray, Int, Long, Long, Long, String, String, Boolean>(ACCOUNT, ACCOUNT.ID, ACCOUNT.HEIGHT, ACCOUNT.CREATION_HEIGHT, ACCOUNT.PUBLIC_KEY, ACCOUNT.KEY_HEIGHT, ACCOUNT.BALANCE,
                                    ACCOUNT.UNCONFIRMED_BALANCE, ACCOUNT.FORGED_BALANCE, ACCOUNT.NAME, ACCOUNT.DESCRIPTION, ACCOUNT.LATEST)
                                    .key(ACCOUNT.ID, ACCOUNT.HEIGHT)
                                    .values(account.id, height, account.creationHeight, account.publicKey, account.keyHeight,
                                            account.balanceNQT, account.unconfirmedBalanceNQT, account.forgedBalanceNQT, account.name, account.description, true)
                    )
                }
                ctx.batch(accountQueries).execute()
            }
        }
    }

    override fun getAssetAccountsCount(assetId: Long): Int {
        return dp.db.useDslContext<Int> { ctx -> ctx.selectCount().from(ACCOUNT_ASSET).where(ACCOUNT_ASSET.ASSET_ID.eq(assetId)).and(ACCOUNT_ASSET.LATEST.isTrue).fetchOne(0, Int::class.javaPrimitiveType) }
    }

    override fun getAccountsWithRewardRecipient(recipientId: Long?): Collection<Account.RewardRecipientAssignment> {
        return rewardRecipientAssignmentTable.getManyBy(getAccountsWithRewardRecipientClause(recipientId!!, dp.blockchain.height + 1), 0, -1)
    }

    override fun getAssets(from: Int, to: Int, id: Long?): Collection<Account.AccountAsset> {
        return accountAssetTable.getManyBy(ACCOUNT_ASSET.ACCOUNT_ID.eq(id), from, to)
    }

    override fun getAssetAccounts(assetId: Long, from: Int, to: Int): Collection<Account.AccountAsset> {
        val sort = mutableListOf<SortField<*>>()
        sort.add(ACCOUNT_ASSET.field("quantity", Long::class.java).desc())
        sort.add(ACCOUNT_ASSET.field("account_id", Long::class.java).asc())
        return accountAssetTable.getManyBy(ACCOUNT_ASSET.ASSET_ID.eq(assetId), from, to, sort)
    }

    override fun getAssetAccounts(assetId: Long, height: Int, from: Int, to: Int): Collection<Account.AccountAsset> {
        if (height < 0) {
            return getAssetAccounts(assetId, from, to)
        }

        val sort = mutableListOf<SortField<*>>()
        sort.add(ACCOUNT_ASSET.field("quantity", Long::class.java).desc())
        sort.add(ACCOUNT_ASSET.field("account_id", Long::class.java).asc())
        return accountAssetTable.getManyBy(ACCOUNT_ASSET.ASSET_ID.eq(assetId), height, from, to, sort)
    }

    override fun setOrVerify(acc: Account, key: ByteArray, height: Int): Boolean {
        when {
            acc.publicKey == null -> {
                if (dp.db.isInTransaction) {
                    acc.publicKey = key
                    acc.keyHeight = -1
                    accountTable.insert(acc)
                }
                return true
            }
            Arrays.equals(acc.publicKey, key) -> return true
            acc.keyHeight == -1 -> {
                if (logger.isInfoEnabled) {
                    logger.info("DUPLICATE KEY!!!")
                    logger.info("Account key for {} was already set to a different one at the same height, current height is {}, rejecting new key", acc.id.toUnsignedString(), height)
                }
                return false
            }
            acc.keyHeight >= height -> {
                logger.info("DUPLICATE KEY!!!")
                if (dp.db.isInTransaction) {
                    if (logger.isInfoEnabled) {
                        logger.info("Changing key for account {} at height {}, was previously set to a different one at height {}", acc.id.toUnsignedString(), height, acc.keyHeight)
                    }
                    acc.publicKey = key
                    acc.keyHeight = height
                    accountTable.insert(acc)
                }
                return true
            }
            logger.isInfoEnabled -> {
                logger.info("DUPLICATE KEY!!!")
                logger.info("Invalid key for account {} at height {}, was already set to a different one at height {}", acc.id.toUnsignedString(), height, acc.keyHeight)
            }
        }
        return false
    }

    internal class SQLAccountAsset(rs: Record) : Account.AccountAsset(rs.get(ACCOUNT_ASSET.ACCOUNT_ID), rs.get(ACCOUNT_ASSET.ASSET_ID), rs.get(ACCOUNT_ASSET.QUANTITY), rs.get(ACCOUNT_ASSET.UNCONFIRMED_QUANTITY), accountAssetDbKeyFactory.newKey(rs.get(ACCOUNT_ASSET.ACCOUNT_ID), rs.get(ACCOUNT_ASSET.ASSET_ID)))

    internal inner class SqlAccount(record: Record) : Account(record.get(ACCOUNT.ID), accountDbKeyFactory.newKey(record.get(ACCOUNT.ID)), record.get(ACCOUNT.CREATION_HEIGHT)) {
        init {
            this.publicKey = record.get(ACCOUNT.PUBLIC_KEY)
            this.keyHeight = record.get(ACCOUNT.KEY_HEIGHT)
            this.balanceNQT = record.get(ACCOUNT.BALANCE)
            this.unconfirmedBalanceNQT = record.get(ACCOUNT.UNCONFIRMED_BALANCE)
            this.forgedBalanceNQT = record.get(ACCOUNT.FORGED_BALANCE)
            this.name = record.get(ACCOUNT.NAME)
            this.description = record.get(ACCOUNT.DESCRIPTION)
        }
    }

    internal inner class SqlRewardRecipientAssignment(record: Record) : Account.RewardRecipientAssignment(record.get(REWARD_RECIP_ASSIGN.ACCOUNT_ID), record.get(REWARD_RECIP_ASSIGN.PREV_RECIP_ID), record.get(REWARD_RECIP_ASSIGN.RECIP_ID), record.get(REWARD_RECIP_ASSIGN.FROM_HEIGHT), rewardRecipientAssignmentDbKeyFactory.newKey(record.get(REWARD_RECIP_ASSIGN.ACCOUNT_ID)))

    companion object {

        private val accountDbKeyFactory = object : DbKey.LongKeyFactory<Account>(ACCOUNT.ID) {
            override fun newKey(account: Account): DbKey {
                return account.nxtKey as DbKey
            }
        }
        private val rewardRecipientAssignmentDbKeyFactory = object : DbKey.LongKeyFactory<Account.RewardRecipientAssignment>(REWARD_RECIP_ASSIGN.ACCOUNT_ID) {
            override fun newKey(assignment: Account.RewardRecipientAssignment): DbKey {
                return assignment.burstKey as DbKey
            }
        }
        private val logger = LoggerFactory.getLogger(SqlAccountStore::class.java)
        private val accountAssetDbKeyFactory = object : DbKey.LinkKeyFactory<Account.AccountAsset>("account_id", "asset_id") {
            override fun newKey(accountAsset: Account.AccountAsset): DbKey {
                return accountAsset.burstKey as DbKey
            }
        }

        private fun getAccountsWithRewardRecipientClause(id: Long, height: Int): Condition {
            return REWARD_RECIP_ASSIGN.RECIP_ID.eq(id).and(REWARD_RECIP_ASSIGN.FROM_HEIGHT.le(height))
        }
    }
}
